cpp/src/arrow/dataset/discovery.cc

```diff
@@ -102,11 +102,10 @@ Result<std::shared_ptr<Dataset>> UnionDatasetFactory::Finish(FinishOptions optio
   return std::shared_ptr<Dataset>(new UnionDataset(options.schema, std::move(children)));
 }
 
-FileSystemDatasetFactory::FileSystemDatasetFactory(
-    std::vector<std::string> paths, std::shared_ptr<fs::FileSystem> filesystem,
-    std::shared_ptr<FileFormat> format, FileSystemFactoryOptions options)
-    : paths_(std::move(paths)),
-      fs_(std::move(filesystem)),
+FileSystemDatasetFactory::FileSystemDatasetFactory(std::vector<FileSource> sources,
+                                                   std::shared_ptr<FileFormat> format,
+                                                   FileSystemFactoryOptions options)
```

Comment:
After this PR, what would be the advantage of that? `FileSystemDatasetFactory.__init__` is the only function which calls `CFileSystemDatasetFactory.MakeFromSources` so I think it would also be the only function which would use `CFileSystemDatasetFactory.MakeFromPaths`. I don't think it'll save any boilerplate

Reply:
OK, if it is only called like this in a single place, it's indeed easy to wrap the files in a list of FileSources at the call site